File: generated_code/GeminiPro/Java/014_SplayTree.java
/**
 * A complete, self-contained Java class implementing an integer set using a Splay Tree.
 * A splay tree is a self-balancing binary search tree with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * The core splay operation. It moves the node with the given key to the root
     * of the tree rooted at {@code node}. If the key is not in the tree, the last
     * accessed node on the search path is moved to the root. This is a top-down splay.
     *
     * @param node The root of the tree (or subtree) to splay.
     * @param key  The key to splay around.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node node, int key) {
        if (node == null) {
            return null;
        }

        // Use a dummy node to simplify linking of left and right subtrees.
        Node dummy = new Node(0);
        Node leftTree = dummy;
        Node rightTree = dummy;

        while (true) {
            if (key < node.key) {
                if (node.left == null) {
                    break;
                }
                // Zig-Zig (left-left) case: Rotate right
                if (key < node.left.key) {
                    Node temp = node.left;
                    node.left = temp.right;
                    temp.right = node;
                    node = temp;
                    if (node.left == null) {
                        break;
                    }
                }
                // Link the current root to the right tree
                rightTree.left = node;
                rightTree = node;
                node = node.left;
            } else if (key > node.key) {
                if (node.right == null) {
                    break;
                }
                // Zig-Zig (right-right) case: Rotate left
                if (key > node.right.key) {
                    Node temp = node.right;
                    node.right = temp.left;
                    temp.left = node;
                    node = temp;
                    if (node.right == null) {
                        break;
                    }
                }
                // Link the current root to the left tree
                leftTree.right = node;
                leftTree = node;
                node = node.right;
            } else {
                // Key found
                break;
            }
        }

        // Reassemble the tree
        leftTree.right = node.left;
        rightTree.left = node.right;
        node.left = dummy.right;
        node.right = dummy.left;

        return node;
    }

    /**
     * Searches for a key in the tree. If the key is found, it is splayed to the
     * root. If not found, the last accessed node on the search path is splayed
     * to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        if (root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the tree. If the key already exists, the node is
     * splayed to the root, and the tree remains unchanged.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key. This brings the node with the key (if it exists)
        // or the would-be parent to the root.
        root = splay(root, key);

        // If the key is already at the root, we're done.
        if (root.key == key) {
            return;
        }

        // Insert the new node as the new root.
        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null; // Detach the old left subtree
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null; // Detach the old right subtree
        }
        root = newNode;
    }

    /**
     * Deletes a key from the tree. If the key is found and deleted, the tree is
     * re-structured by splaying.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree around the key.
        root = splay(root, key);

        // If the key is not at the root after splaying, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // The node to delete is now the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Find the maximum element in the left subtree.
            Node maxNode = leftSubtree;
            while (maxNode.right != null) {
                maxNode = maxNode.right;
            }
            // Splay the left subtree around its maximum element. This makes the max
            // element the root of the left subtree, and it will have no right child.
            leftSubtree = splay(leftSubtree, maxNode.key);

            // Attach the original right subtree as the right child of the new root.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/014_SplayTree.py
import sys

class SplayTree:
    """
    A self-contained Python class that implements a Splay Tree.
    This tree acts as a dictionary-like set for storing unique integer keys.
    """

    class _Node:
        """
        A node in the Splay Tree.
        """
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """
        Initializes an empty Splay Tree.
        """
        self.root = None

    def _left_rotate(self, x):
        """
        Performs a left rotation on node x.
        """
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """
        Performs a right rotation on node x.
        """
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent

        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
            
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splaying operation on node x, moving it to the root.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:
                # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:
                # Zig-Zig case (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:
                # Zig-Zig case (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:
                # Zig-Zag case (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:
                # Zig-Zag case (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.
        
        If the key is found, the corresponding node is splayed to the root and
        True is returned.
        
        If the key is not found, the last accessed node (the parent of where
        the key would be) is splayed to the root and False is returned.
        
        Args:
            key: The integer key to search for.
            
        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        last_node = None
        current = self.root
        while current:
            last_node = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # Key not found, splay the last visited node if it exists
        if last_node:
            self._splay(last_node)
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.
        
        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.
        
        Args:
            key: The integer key to insert.
        """
        # Find the parent for the new node, similar to BST insertion
        parent = None
        current = self.root
        while current:
            parent = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay the node and return
                self._splay(current)
                return

        # Create and link the new node
        new_node = self._Node(key)
        new_node.parent = parent
        
        if not parent:
            self.root = new_node
        elif new_node.key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
            
        # Splay the newly inserted node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.
        
        The tree is first searched for the key, which brings the node (or its
        closest neighbor) to the root. If the key is found at the root, it is
        deleted and the remaining subtrees are joined.
        
        Args:
            key: The integer key to delete.
        """
        if not self.root:
            return

        # Search for the key. This will splay the node or its parent to the root.
        self.search(key)

        # If the key is not at the root after searching, it was not in the tree.
        if self.root.key != key:
            return

        # The node to delete is now at the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left subtree, the right subtree becomes the new tree.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # If there is a left subtree, we need to join it with the right one.
            # 1. Disconnect the left subtree.
            left_subtree.parent = None
            
            # 2. Find the maximum element in the left subtree.
            max_in_left = left_subtree
            while max_in_left.right:
                max_in_left = max_in_left.right
            
            # 3. Splay this maximum element to the root of the left subtree.
            #    After this, it will have no right child.
            #    We temporarily treat the left subtree as the main tree for splaying.
            self.root = left_subtree
            self._splay(max_in_left)
            
            # 4. Attach the original right subtree as the right child of the new root.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root